lib/Remote/JsonWireCompat.php

```diff
@@ -0,0 +1,75 @@
+<?php
+// Copyright 2004-present Facebook. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+namespace Facebook\WebDriver\Remote;
+
+use Facebook\WebDriver\WebDriverBy;
+
+/**
+ * Compatibility layer between W3C's WebDriver and the legacy JsonWire protocol.
+ *
+ * @internal
+ */
+abstract class JsonWireCompat
+{
+    /**
+     * Element identifier defined in the W3C's WebDriver protocol.
+     *
+     * @see https://w3c.github.io/webdriver/webdriver-spec.html#elements
+     */
+    const WEB_DRIVER_ELEMENT_IDENTIFIER = 'element-6066-11e4-a52e-4f735466cecf';
+
+    public static function getElement(array $rawElement)
+    {
+        if (array_key_exists(self::WEB_DRIVER_ELEMENT_IDENTIFIER, $rawElement)) {
+            // W3C's WebDriver
+            return $rawElement[self::WEB_DRIVER_ELEMENT_IDENTIFIER];
+        }
+
+        // Legacy JsonWire
+        return $rawElement['ELEMENT'];
+    }
+
+    /**
+     * @param WebDriverBy $by
+     * @param bool $w3cCompliance
+     * @return array
+     */
+    public static function getUsing(WebDriverBy $by, $w3cCompliance)
+    {
+        $mechanism = $by->getMechanism();
+        $value = $by->getValue();
+
+        if ($w3cCompliance) {
+            switch ($mechanism) {
+                // Convert to CSS selectors
+                case 'class name':
+                    $mechanism = 'css selector';
+                    $value = ".$value";
```

Comment:
should it escape special chars (brackets, etc...). I don't know how the JsonWire protocol supports such cases. If it is able to work with them, they need to be escaped when building a CSS selector

Reply:
the same is true for other cases

Reply:
It's not possible in CSS to escape special chars in identifiers: https://mathiasbynens.be/notes/css-escapes
The only thing we can do is throwing an error.

Reply:
the article you linked is precisely explaining how to escape special chars to build selectors (see examples in https://mathiasbynens.be/notes/css-escapes#trailing-whitespace for instance). So I don't agree with you

Reply:
Is this valid according to the spec?

> In CSS, identifiers (including element names, classes, and IDs in selectors) can contain only the characters [a-zA-Z0-9] and ISO 10646 characters U+00A0 and higher, plus the hyphen (-) and the underscore (_); they cannot start with a digit, two hyphens, or a hyphen followed by a digit. Identifiers can also contain escaped characters and any ISO 10646 character as a numeric code (see next item). For instance, the identifier "B&W?" may be written as "B\&W\?" or "B\26 W\3F".
https://www.w3.org/TR/CSS21/syndata.html#value-def-identifier

Reply:
> Identifiers can also contain escaped characters and any ISO 10646 character as a numeric code

When escaping them, you can put what you want

Reply:
Escaping is now implemented